bser/encode.go

```diff
@@ -176,6 +179,36 @@ func encode(buf []byte, d interface{}) ([]byte, error) {
 	}
 }
 
+func encodeTemplate(buf []byte, r reflect.Value) ([]byte, error) {
+	elem := r.Type().Elem()
+	fieldNames := make([]string, elem.NumField())
+	b, err := encode(nil, r.Len())
+	if err != nil {
+		return nil, err
+	}
+	for i := 0; i < r.Len(); i++ {
+		s := r.Index(i)
+		for j := 0; j < s.NumField(); j++ {
+			if i == 0 {
```

Comment:
do we need to worry about unexpected fields?

Reply:
@jonasi could you clarify what you mean by that? Are you referring to a situation where different structs in the array might have different fields (e.g. `[{A: 1, B: 2}, {C: 3, D:4}]`) or is there something else?

I don't think it is possible to have an array of different structs in go, unless maybe it is an array of interfaces, [in which case it wouldn't pass](https://play.golang.org/p/TfsJYHIxbio) the `if r.Type().Elem().Kind() == reflect.Struct`  check we have prior to calling `encodeTemplate(buf, r)`

Reply:
@ShawnROGrady I mean for a type like `struct { A string; b string }`. Do we have to worry about `b`?

Reply:
@jonasi good catch! Turns out we weren't handling this correctly in the non-template case either which resulted in a panic when a struct had an unexported field. Added a new `exportedStructFields()` helper to correctly handle this

Reply:
@jonasi moved those changes to a separate PR: https://github.com/jonasi/watchman/pull/21

Reply:
haha sorry - just realized I got autocorrected to `unexpected`.  obviously meant `unexported` in my first post...